### PR TITLE
added DAGMCUniverse.bounding_region

### PR DESCRIPTION
**Pull request by shimwell**
As discussed with @pshriwise in issue #2104 and building on the work from PR #2111 this is another small PR that adds the ability to automatically generate a bounding region around a DAGMC geometry.

This adds the ```DAGMCUniverse.bounding_region()``` method

This would be useful for creating a spherical or box shaped vacuum surface to bound the DAGMC geometry or other uses (boundary type is user controllable)

This is based on some code from another micro package I made a while back :recycle: [here](https://github.com/fusion-energy/openmc-dagmc-wrapper/blob/9c4485d4f40e7df2bac5333be68fe6d4e0caa570/openmc_dagmc_wrapper/Geometry.py#L114-L153) and [here](https://github.com/fusion-energy/openmc-dagmc-wrapper/blob/9c4485d4f40e7df2bac5333be68fe6d4e0caa570/openmc_dagmc_wrapper/Geometry.py#L50-L55)

I've added some simple tests to the DAGMC universe test and type checking

Current method of bringing a graveyard free DAGMC geometry file into to OpenMC
```python
# makes use of the dagmc geometry
dag_univ = openmc.DAGMCUniverse("dagmc.h5m")

# creates an edge of universe boundary at a large radius
vac_surf = openmc.Sphere(r=10000, surface_id=9999, boundary_type="vacuum")

# specifies the region as below the universe boundary
region = -vac_surf

# creates a cell from the region and fills the cell with the dagmc geometry
containing_cell = openmc.Cell(cell_id=9999, region=region, fill=dag_univ)

model = openmc.Model()
model.geometry = openmc.Geometry(root=[containing_cell])
```

method possible with this PR of bringing a DAGMC geometry file into to OpenMC
```python
# loads up dagmc file and automatically bounds the geometry with a vacuum surface
bound_dag_univ= openmc.DAGMCUniverse("dagmc.h5m").bounded_universe()

model = openmc.Model()
model.geometry = openmc.Geometry(root=bound_dag_univ)
```

### CODE REVIEW

File: openmc/universe.py [hunk 1]
```diff
@@ -713,6 +716,57 @@ def create_xml_subelement(self, xml_element, memo=None):
...
 
+    def bounding_region(self, bounded_type='box', boundary_type='vacuum'):
+        """Creates a either a spherical or box shaped bounding region around
+        the DAGMC geometry.
+        Parameters
+        ----------
+        bounded_type : str
+            The type of bounding surface(s) to use when constructing the region.
+            Options include a single spherical surface (sphere) or a rectangle
+            made from siz planes (box).
```
Comment (pshriwise) [inline openmc/universe.py:727]:
```suggestion
            made from six planes (box).
```

File: openmc/universe.py [hunk 2]
```diff
@@ -713,6 +716,57 @@ def create_xml_subelement(self, xml_element, memo=None):
...
+        openmc.Region
+            Region instance
+        """
+
+        check_type('boundary type', boundary_type, str)
+        check_value('boundary type', boundary_type, _BOUNDARY_TYPES)
+
+        bounding_box = self.bounding_box
+
+        if bounded_type == 'sphere':
```
Comment (pshriwise) [inline openmc/universe.py:743]: Can we add type/value checks for `bounded_type`?

File: openmc/universe.py [hunk 3]
```diff
@@ -713,6 +716,57 @@ def create_xml_subelement(self, xml_element, memo=None):
...
+            upper_x = openmc.XPlane(bounding_box[1][0], boundary_type=boundary_type)
+            lower_y = openmc.YPlane(bounding_box[0][1], boundary_type=boundary_type)
+            upper_y = openmc.YPlane(bounding_box[1][1], boundary_type=boundary_type)
+            lower_z = openmc.ZPlane(bounding_box[0][2], boundary_type=boundary_type)
+            upper_z = openmc.ZPlane(bounding_box[1][2], boundary_type=boundary_type)
+
+            return +lower_x & -upper_x & +lower_y & -upper_y & +lower_z & -upper_z
+
+    def bounded_universe(self, **kwargs):
+        """Returns an openmc.Universe filled with the DAGMCUniverse and bounded
```
Comment (pshriwise) [inline openmc/universe.py:763]:
```suggestion
        """Returns an openmc.Universe filled with this DAGMCUniverse and bounded
```

File: openmc/universe.py (hunk 2 above)
Comment (shimwell) [reply openmc/universe.py:745]: good idea, I've just added type and value checks in commit https://github.com/openmc-dev/openmc/pull/2118/commits/839b862d5645be785fd1f79dc6c3ba3f19e39b9e

**Comment by shimwell** [comment]
I think I've missed some logic handling as I'm now getting this stack track while testing

TypeError                                 Traceback (most recent call last)
E           Input In [6], in <cell line: 3>()
E                 1 import openmc
E                 2 bound_dag_univ = openmc.DAGMCUniverse(filename='dagmc.h5m').bounded_universe()
E           ----> 3 geometry = openmc.Geometry(root=[bound_dag_univ])
E           
E           File /opt/conda/lib/python3.8/site-packages/openmc/geometry.py:40, in Geometry.__init__(self, root)
E                38 univ = openmc.Universe()
E                39 for cell in root:
E           ---> 40     univ.add_cell(cell)
E                41 self._root_universe = univ
E           
E           File /opt/conda/lib/python3.8/site-packages/openmc/universe.py:379, in Universe.add_cell(self, cell)
E               376 if not isinstance(cell, openmc.Cell):
E               377     msg = f'Unable to add a Cell to Universe ID="{self._id}" since ' \
E               378           f'"{cell}" is not a Cell'
E           --> 379     raise TypeError(msg)
E               381 cell_id = cell.id
E               383 if cell_id not in self._cells:
E           
E           TypeError: Unable to add a Cell to Universe ID="3" since "Universe
E           	ID             =	2
E           	Name           =	
E           	Geom           =	CSG
E           	Cells          =	[1]
E           " is not a Cell

**Comment by pshriwise** [comment]
> geometry = openmc.Geometry(root=[bound_dag_univ])

I think this line should be 

```python
 geometry = openmc.Geometry(root=bound_dag_univ)
```

That constructor either takes a single root universe or a list of cells to add to a root universe, but not a list of unvierses.

**Comment by shimwell** [comment]
to avoid a ```RuntimeError: Cell ID 1 exists in both DAGMC Universe 1 and the CSG geometry``` error I've set the ```auto_geom_id``` to True by default when using ```bounding_universe``` method.

File: openmc/universe.py [hunk 4]
```diff
@@ -713,6 +716,74 @@ def create_xml_subelement(self, xml_element, memo=None):
...
+        check_type('bounded type', bounded_type, str)
+        check_value('bounded type', bounded_type, ('box', 'sphere'))
+
+        bounding_box = self.bounding_box
+
+        if bounded_type == 'sphere':
+            import math
+            bounding_box_center = (bounding_box[0] + bounding_box[1])/2
+            radius = math.dist(bounding_box[0], bounding_box[1])
+            bounding_surface = openmc.Sphere(x0=bounding_box_center[0], y0=bounding_box_center[1], z0=bounding_box_center[2], r=radius)
```
Comment (pshriwise) [inline openmc/universe.py:749]: From the tests it looks like we need to pass the `boundary_type` through to the sphere.

**Comment by pshriwise** [comment]
Looks like we'll need to update the inputs for that DAGMC universes test since we're creating additional surfaces now.

**Comment by shimwell** [comment]
I'm still a bit cautious on the ID front, do you think we have fully guarded against this error that might occur if the cell IDs are not done correctly

```RuntimeError: Cell ID 1 exists in both DAGMC Universe 1 and the CSG geometry.```

**Comment by pshriwise** [comment]
> I'm still a bit cautious on the ID front, do you think we have fully guarded against this error that might occur if the cell IDs are not done correctly

I believe so, yes. This is currently handled in the C++ code. Is there another place you're running into this error?

File: openmc/universe.py [hunk 5]
```diff
@@ -713,6 +716,72 @@ def create_xml_subelement(self, xml_element, memo=None):
...
+        DAGMCUniverse.bounding_region().
+
+        Returns
+        -------
+        openmc.Universe
+            Universe instance
+        """
+
+        bounding_cell = openmc.Cell(fill=self, region=self.bounding_region(**kwargs))
+        return openmc.Universe(cells=[bounding_cell])
```
Comment (shimwell) [inline openmc/universe.py:783]: @pshriwise what do you think of this idea, we specify the ID number for the bounding cell, then if the user normally won't need to pass in the auto_id=True argument quite as often.

```suggestion
    def bounded_universe(self, bounding_cell_id=10000, **kwargs):
        """Returns an openmc.Universe filled with this DAGMCUniverse and bounded
        with a cell. Defaults to a box cell with a vacuum surface however this
        can be changed using the kwargs which are passed directly to
        DAGMCUniverse.bounding_region().

        Parameters
        ----------
        bounding_cell_id : int
            The cell ID number to use for the bounding cell, defaults to 1000 to reduce
            the chance of overlapping ID numbers with the DAGMC geometry.

        Returns
        -------
        openmc.Universe
            Universe instance
        """

        bounding_cell = openmc.Cell(fill=self, id=bounding_cell_id, region=self.bounding_region(**kwargs))
        return openmc.Universe(cells=[bounding_cell])
```

File: src/dagmc.cpp [hunk 6]
```diff
@@ -183,7 +183,9 @@ void DAGUniverse::init_geometry()
     } else {
       warning(fmt::format("DAGMC Cell IDs: {}", dagmc_ids_for_dim(3)));
       fatal_error(fmt::format("Cell ID {} exists in both DAGMC Universe {} "
-                              "and the CSG geometry.",
+                              "and the CSG geometry. Setting auto_geom_ids
+                              to True when initiating the DAGMC Universe may
+                              resolve this issue",
```
Comment (shimwell) [inline src/dagmc.cpp:188]:
```suggestion
                              "and the CSG geometry. Setting auto_geom_ids "
                              "to True when initiating the DAGMC Universe may "
                              "resolve this issue",
```

File: openmc/universe.py (hunk 5 above)
Comment (pshriwise) [reply openmc/universe.py:783]: Yeah, I think that should be okay. 👍🏻

File: openmc/universe.py [hunk 7]
```diff
@@ -713,6 +716,78 @@ def create_xml_subelement(self, xml_element, memo=None):
...
+            The cell ID number to use for the bounding cell, defaults to 1000 to reduce
+            the chance of overlapping ID numbers with the DAGMC geometry.
+
+        Returns
+        -------
+        openmc.Universe
+            Universe instance
+        """
+
+        bounding_cell = openmc.Cell(fill=self, id=bounding_cell_id, region=self.bounding_region(**kwargs))
```
Comment (shimwell) [inline openmc/universe.py:788]:
```suggestion
        bounding_cell = openmc.Cell(fill=self, cell_id =bounding_cell_id, region=self.bounding_region(**kwargs))
```

**Review by pshriwise — APPROVED** [review]
Looks good @shimwell! Thanks for this! Hopefully it will make using DAGMC models w/ OpenMC that much easier. 👍🏻

**Comment by pshriwise** [comment]
Closes #2104

**Pull request merged**